src/persist/src/indexed/columnar.rs

```diff
@@ -523,6 +524,22 @@ impl ColumnarRecords {
     }
 }
 
+/// An "extension" to [`ColumnarRecords`] that duplicates the "key" (`K`) and "val" (`V`) columns
+/// as structured Arrow data.
+///
+/// [`ColumnarRecords`] stores the key and value columns as binary blobs encoded with the [`Codec`]
+/// trait. We're migrating to instead store the key and value columns as structured Parquet data,
+/// which we interface with via Arrow.
+///
+/// [`Codec`]: mz_persist_types::Codec
+#[derive(Debug, Clone)]
+pub struct ColumnarRecordsStructuredExt {
+    /// The structured `k` column.
+    pub key: Option<StructArray>,
+    /// The structured `v` column.
+    pub val: Option<StructArray>,
+}
```

Comment:
Are there any circumstances in which one of these should be set and the other is not? If not, thoughts on moving the option up one level?

Reply:
AFAIU [`arrow2` doesn't allow empty struct arrays](https://github.com/MaterializeInc/materialize/blob/main/src/persist-types/src/part.rs#L63-L67), so if `key` or `val` are ever empty (which `val` almost always is) it gets modeled as `None` here.